Flaky test_geth_request_pruned_data_raises_an_exception

```
______________ test_geth_request_pruned_data_raises_an_exception _______________

deploy_client = <JSONRPCClient node:0x36B72cB4FF2b725b11d9028e17c879d4Fa1Cae83 nonce:64>
web3 = <web3.main.Web3 object at 0x7f033eca2e10>

    def test_geth_request_pruned_data_raises_an_exception(deploy_client, web3):
        """ Interacting with an old block identifier with a pruning client throws. """
        contract_proxy, _ = deploy_rpc_test_contract(deploy_client, "RpcWithStorageTest")
        iterations = 1
    
        def send_transaction():
            check_block = deploy_client.get_checking_block()
            startgas = contract_proxy.estimate_gas(check_block, "waste_storage", iterations)
            startgas = safe_gas_limit(startgas)
            transaction = contract_proxy.transact("waste_storage", startgas, iterations)
            deploy_client.poll(transaction)
            return deploy_client.get_transaction_receipt(transaction)
    
        first_receipt = send_transaction()
        pruned_block_number = first_receipt["blockNumber"]
    
        # geth keeps the latest 128 blocks before pruning. Unfortunately, this can
        # not be configured to speed this test up.
        non_pruned_blocks = 128
        while web3.eth.blockNumber < pruned_block_number + non_pruned_blocks + 1:
            send_transaction()
    
        with pytest.raises(ValueError):
>           contract_proxy.contract.functions.const().call(block_identifier=pruned_block_number)
E           Failed: DID NOT RAISE <class 'ValueError'>

raiden/tests/integration/rpc/assumptions/test_geth_rpc_assumptions.py:32: Failed
```

Seen in https://circleci.com/gh/raiden-network/raiden/69706#tests/containers/19

I've already compared the raiden log from CI (failing) and local tests (succeeding) and except for time stamps and paths, they are identical. I also reproduced the same test execution order locally to exclude side effects of previous tests as reason. I the geth log on CI, the expected messages of the form
```
Served eth_call                          conn=127.0.0.1:59392 reqid=650 t=213.5µs err="missing trie node 07502fc5570b50a659c36654ab7a21ece3674dc793a45305f09abc00a92f3e4b (path )"
```
are missing.

Maybe geth promises to keep 128 blocks, but might keep slightly more under certain circumstances (e.g. the next purging run did not happen yet). If we don’t find a better explanation, we could mine a few blocks more and see if that solves the flakyness.

--- a/raiden/tests/integration/rpc/assumptions/test_geth_rpc_assumptions.py
+++ b/raiden/tests/integration/rpc/assumptions/test_geth_rpc_assumptions.py
@@ -6,7 +6,6 @@ from raiden.utils import safe_gas_limit
 pytestmark = pytest.mark.usefixtures("skip_if_not_geth")
 
 
-@pytest.mark.skip(reason="Flaky, see https://github.com/raiden-network/raiden/issues/4545")
 def test_geth_request_pruned_data_raises_an_exception(deploy_client, web3):
     """ Interacting with an old block identifier with a pruning client throws. """
     contract_proxy, _ = deploy_rpc_test_contract(deploy_client, "RpcWithStorageTest")
@@ -26,7 +25,14 @@ def test_geth_request_pruned_data_raises_an_exception(deploy_client, web3):
     # geth keeps the latest 128 blocks before pruning. Unfortunately, this can
     # not be configured to speed this test up.
     non_pruned_blocks = 128
-    while web3.eth.blockNumber < pruned_block_number + non_pruned_blocks + 1:
+    # According to the geth devs if we run a PoA chain (clique, dev mode) and
+    # HEAD-127 doesn't contain any transactions, then the state of HEAD-127 will
+    # be the same as HEAD-128, so it will still be referenced and not deleted.
+    # To avoid this and make sure the pruned_block_number is pruned let's mine
+    # buffer period more blocks while at the same time sending transactions
+    buffer_period_blocks = 5
+    target_block = pruned_block_number + non_pruned_blocks + buffer_period_blocks + 1
+    while web3.eth.blockNumber <= target_block:
         send_transaction()
 
     with pytest.raises(ValueError):

--- a/raiden/tests/integration/rpc/assumptions/test_geth_rpc_assumptions.py
+++ b/raiden/tests/integration/rpc/assumptions/test_geth_rpc_assumptions.py
@@ -1,4 +1,6 @@
+import gevent
 import pytest
+
 
 from raiden.tests.utils.smartcontracts import deploy_rpc_test_contract
 from raiden.utils import safe_gas_limit
@@ -20,21 +22,20 @@ def test_geth_request_pruned_data_raises_an_exception(deploy_client, web3):
         return deploy_client.get_transaction_receipt(transaction)
 
     first_receipt = send_transaction()
-    pruned_block_number = first_receipt["blockNumber"]
+    mined_block_number = first_receipt["blockNumber"]
+
+    while mined_block_number + 127 > web3.eth.blockNumber:
+        gevent.sleep(0.5)
 
     # geth keeps the latest 128 blocks before pruning. Unfortunately, this can
     # not be configured to speed this test up.
-    non_pruned_blocks = 128
     # According to the geth devs if we run a PoA chain (clique, dev mode) and
     # HEAD-127 doesn't contain any transactions, then the state of HEAD-127 will
     # be the same as HEAD-128, so it will still be referenced and not deleted.
-    # To avoid this and make sure the pruned_block_number is pruned let's mine
-    # buffer period more blocks while at the same time sending transactions
-    buffer_period_blocks = 5
-    target_block = pruned_block_number + non_pruned_blocks + buffer_period_blocks + 1
-    while web3.eth.blockNumber <= target_block:
-        send_transaction()
+    # So for this test we mine a transaction, wait for mined + 127 block and then
+    # query the previous block which should be pruned
 
+    pruned_block_number = mined_block_number - 1
     with pytest.raises(ValueError):
         contract_proxy.contract.functions.const().call(block_identifier=pruned_block_number)
 

